### PR TITLE
Interpreter: Print BlockInputStream

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: ref #4609 

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Interpreters/executeQuery.cpp [hunk 1]
```diff
@@ -187,7 +187,7 @@ std::tuple<ASTPtr, BlockIO> executeQueryImpl(
     }
     catch (...)
     {
-        if (!internal)
+        if (!internal && !context.getDAGContext()->isTest())
```
Comment (ywqzzy) [inline dbms/src/Interpreters/executeQuery.cpp:190]: I will remove the !context.getDAGContext()->isTest() after code review.

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 2]
```diff
@@ -944,7 +968,7 @@ BlockInputStreams DAGQueryBlockInterpreter::execute()
     executeImpl(pipeline);
     if (!pipeline.streams_with_non_joined_data.empty())
     {
-        executeUnion(pipeline, max_streams, log);
+        executeUnion(pipeline, max_streams, log, false, "final union");
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:971]:
```suggestion
        executeUnion(pipeline, max_streams, log, false, "final union for non_joined_data");
```
```suggestion
        executeUnion(pipeline, max_streams, log, false, "final union");
```

Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:971]:
```suggestion
        executeUnion(pipeline, max_streams, log, false, "final union for non_joined_data");
```
```suggestion
        executeUnion(pipeline, max_streams, log, false, "final union");
```

File: dbms/src/Flash/Coprocessor/InterpreterUtils.cpp [hunk 3]
```diff
@@ -46,13 +47,22 @@ BlockInputStreamPtr combinedNonJoinedDataStream(
 {
     BlockInputStreamPtr ret = nullptr;
     if (pipeline.streams_with_non_joined_data.size() == 1)
+    {
         ret = pipeline.streams_with_non_joined_data.at(0);
+        ret->setExtraInfo("non joined data");
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/InterpreterUtils.cpp:52]: No need to set it again here

Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/InterpreterUtils.cpp:52]: No need to set it again here

File: dbms/src/Flash/tests/gtest_interpreter.cpp [hunk 4]
```diff
@@ -47,18 +47,18 @@ try
...
+  Expression: <final projection>
+   MergeSorting, limit = 10
+    Union: <for order>
+     PartialSorting x 10: limit = 10
+      Expression: <before order and select>
+       Filter: <execute having>: {equals->Nullable(UInt8)}
+        SharedQuery: <restore concurrency>
+         ParallelAggregating, max_threads: 10, final: true
+          Expression x 10: <before aggregation>
+           Filter: <execute where>: {equals->Nullable(UInt8)}
```
Comment (SeaRise) [inline dbms/src/Flash/tests/gtest_interpreter.cpp:61]: Does `{equals->Nullable(UInt8)}` make sense in a unit test?

Comment (SeaRise) [inline dbms/src/Flash/tests/gtest_interpreter.cpp:61]: Does `{equals->Nullable(UInt8)}` make sense in a unit test?

File: dbms/src/Flash/tests/gtest_interpreter.cpp [hunk 5]
```diff
@@ -47,18 +47,18 @@ try
...
-    Union
-     PartialSorting x 10
-      Expression
-       Filter
-        SharedQuery
-         ParallelAggregating
-          Expression x 10
-           Filter
+Union: <for mpp>
+ SharedQuery x 10: <restore concurrency>
```
[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Flash/tests/gtest_interpreter.cpp:51]: and `non_joined_streams.empty()`?
Do it in a new pr is better to me.

[the remark this answers is not in the document]
Comment (ywqzzy) [reply dbms/src/Flash/tests/gtest_interpreter.cpp:51]: > and `non_joined_streams.empty()`? Do it in a new pr is better to me.

So do I, I will create a new pr.

File: dbms/src/DataStreams/IBlockInputStream.h [hunk 6]
```diff
@@ -176,6 +177,8 @@ class IBlockInputStream : private boost::noncopyable
         }
     }
 
+    virtual void appendInfo(FmtBuffer & buffer [[maybe_unused]]) const {};
```
Comment (SeaRise) [inline dbms/src/DataStreams/IBlockInputStream.h:180]:
```suggestion
    virtual void appendInfo(FmtBuffer & /*buffer*/) const {};
```

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 7]
```diff
@@ -254,7 +257,7 @@ void getJoinKeyTypes(const tipb::Join & join, DataTypes & key_types)
     }
 }
 
-void DAGQueryBlockInterpreter::handleJoin(const tipb::Join & join, DAGPipeline & pipeline, SubqueryForSet & right_query)
+void DAGQueryBlockInterpreter::handleJoin(const tipb::Join & join, DAGPipeline & pipeline, SubqueryForSet & right_query, const String & executor_id)
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:260]: how about using `query_block.source_name` directly in `handleJoin`?

File: dbms/src/Interpreters/ExpressionActions.cpp [hunk 8]
```diff
@@ -931,8 +929,8 @@ std::string ExpressionActions::dumpActions() const
 
     ss << "\noutput:\n";
     NamesAndTypesList output_columns = sample_block.getNamesAndTypesList();
-    for (NamesAndTypesList::const_iterator it = output_columns.begin(); it != output_columns.end(); ++it)
-        ss << it->name << " " << it->type->getName() << "\n";
+    for (auto & output_column : output_columns)
```
Comment (SeaRise) [inline dbms/src/Interpreters/ExpressionActions.cpp:932]:
```suggestion
    for (const auto & output_column : output_columns)
```

File: dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp [hunk 9]
```diff
@@ -412,6 +414,7 @@ void DAGStorageInterpreter::executeCastAfterTableScan(
         {
             auto & stream = pipeline.streams[i++];
             stream = std::make_shared<ExpressionBlockInputStream>(stream, extra_cast, log->identifier());
+            stream->setExtraInfo("cast after tableScan");
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp:417]:
```suggestion
            stream->setExtraInfo("cast after local tableScan");
```

File: dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp [hunk 10]
```diff
@@ -424,6 +427,7 @@ void DAGStorageInterpreter::executeCastAfterTableScan(
             {
                 auto & stream = pipeline.streams[i++];
                 stream = std::make_shared<ExpressionBlockInputStream>(stream, project_for_cop_read, log->identifier());
+                stream->setExtraInfo("cast after tableScan");
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp:430]:
```suggestion
                stream->setExtraInfo("cast after remote tableScan");
```

File: dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp [hunk 11]
```diff
@@ -373,8 +373,10 @@ void DAGStorageInterpreter::executePushedDownFilter(
     {
         auto & stream = pipeline.streams[i];
         stream = std::make_shared<FilterBlockInputStream>(stream, before_where, filter_column_name, log->identifier());
+        stream->setExtraInfo("push down filter");
         // after filter, do project action to keep the schema of local streams and remote streams the same.
         stream = std::make_shared<ExpressionBlockInputStream>(stream, project_after_where, log->identifier());
+        stream->setExtraInfo("push down filter");
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp:379]:
```suggestion
        stream->setExtraInfo("projection after push down filter");
```

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 12]
```diff
@@ -258,20 +258,22 @@ void DAGQueryBlockInterpreter::handleJoin(const tipb::Join & join, DAGPipeline &
...
 
     /// build side streams
-    executeExpression(build_pipeline, build_side_prepare_actions);
+    executeExpression(build_pipeline, build_side_prepare_actions, "Append Join key and join Filters");
     // add a HashJoinBuildBlockInputStream to build a shared hash table
     auto get_concurrency_build_index = JoinInterpreterHelper::concurrencyBuildIndexGenerator(join_build_concurrency);
     build_pipeline.transform([&](auto & stream) {
         stream = std::make_shared<HashJoinBuildBlockInputStream>(stream, join_ptr, get_concurrency_build_index(), log->identifier());
+        stream->setExtraInfo(
+            fmt::format("join build, root_executor_id = {}", dagContext().getJoinExecuteInfoMap()[query_block.source_name].build_side_root_executor_id));
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:267]:
```suggestion
            fmt::format("join build, build_side_root_executor_id = {}", dagContext().getJoinExecuteInfoMap()[query_block.source_name].build_side_root_executor_id));
```

File: dbms/src/DataStreams/FilterBlockInputStream.cpp [hunk 13]
```diff
@@ -16,6 +16,7 @@
 #include <Columns/ColumnsCommon.h>
 #include <Columns/ColumnsNumber.h>
 #include <Columns/FilterDescription.h>
+#include <Common/FmtUtils.h>
```
Comment (SeaRise) [inline dbms/src/DataStreams/FilterBlockInputStream.cpp:19]: useless import?

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.h [hunk 14]
```diff
@@ -66,8 +66,8 @@ class DAGQueryBlockInterpreter
     void handleProjection(DAGPipeline & pipeline, const tipb::Projection & projection);
     void handleWindow(DAGPipeline & pipeline, const tipb::Window & window);
     void handleWindowOrder(DAGPipeline & pipeline, const tipb::Sort & window_sort);
-    void executeWhere(DAGPipeline & pipeline, const ExpressionActionsPtr & expressionActionsPtr, String & filter_column);
-    void executeExpression(DAGPipeline & pipeline, const ExpressionActionsPtr & expressionActionsPtr);
+    void executeWhere(DAGPipeline & pipeline, const ExpressionActionsPtr & expressionActionsPtr, String & filter_column, String extra_info = "");
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.h:69]:
```suggestion
    void executeWhere(DAGPipeline & pipeline, const ExpressionActionsPtr & expressionActionsPtr, String & filter_column, const String & extra_info = "");
```

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.h [hunk 15]
```diff
@@ -66,8 +66,8 @@ class DAGQueryBlockInterpreter
...
+    void executeExpression(DAGPipeline & pipeline, const ExpressionActionsPtr & expressionActionsPtr, String extra_info = "");
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.h:70]: ditto

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.h [hunk 16]
```diff
@@ -82,7 +82,7 @@ class DAGQueryBlockInterpreter
         const TiDB::TiDBCollators & collators,
         AggregateDescriptions & aggregate_descriptions,
         bool is_final_agg);
-    void executeProject(DAGPipeline & pipeline, NamesWithAliases & project_cols);
+    void executeProject(DAGPipeline & pipeline, NamesWithAliases & project_cols, String extra_info = "");
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.h:85]: ditto

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 17]
```diff
@@ -258,20 +258,22 @@ void DAGQueryBlockInterpreter::handleJoin(const tipb::Join & join, DAGPipeline &
     size_t join_build_concurrency = settings.join_concurrent_build ? std::min(max_streams, build_pipeline.streams.size()) : 1;
 
     /// build side streams
-    executeExpression(build_pipeline, build_side_prepare_actions);
+    executeExpression(build_pipeline, build_side_prepare_actions, "Append Join key and join Filters");
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:261]: build side prepare actions?
or
Append Join key and join Filters for build side?

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 18]
```diff
@@ -258,20 +258,22 @@ void DAGQueryBlockInterpreter::handleJoin(const tipb::Join & join, DAGPipeline &
...
-    executeUnion(build_pipeline, max_streams, log, /*ignore_block=*/true);
+    executeUnion(build_pipeline, max_streams, log, /*ignore_block=*/true, "for join");
 
     right_query.source = build_pipeline.firstStream();
     right_query.join = join_ptr;
     join_ptr->init(right_query.source->getHeader(), join_build_concurrency);
 
     /// probe side streams
-    executeExpression(probe_pipeline, probe_side_prepare_actions);
+    executeExpression(probe_pipeline, probe_side_prepare_actions, "probe side prepare actions");
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:276]: or
Append Join key and join Filters for probe side?

**Review by SeaRise — COMMENTED** [review]
other LGTM

File: dbms/src/Flash/Coprocessor/InterpreterDAG.cpp [hunk 19]
```diff
@@ -26,7 +26,7 @@ InterpreterDAG::InterpreterDAG(Context & context_, const DAGQuerySource & dag_)
     , dag(dag_)
 {
     const Settings & settings = context.getSettingsRef();
-    if (dagContext().isBatchCop() || dagContext().isMPPTask())
+    if (dagContext().isBatchCop() || (dagContext().isMPPTask() && !dagContext().isTest()))
         max_streams = settings.max_threads;
     else if (dagContext().isTest())
         max_streams = dagContext().initialize_concurrency;
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/InterpreterDAG.cpp:32]:
```suggestion
    if (dagContext().isTest())
        max_streams = dagContext().initialize_concurrency;
    else if (dagContext().isBatchCop() || dagContext().isMPPTask())
        max_streams = settings.max_threads;
```

File: dbms/src/Flash/Coprocessor/InterpreterDAG.cpp [hunk 20]
```diff
@@ -85,9 +85,9 @@ BlockIO InterpreterDAG::execute()
     /// add union to run in parallel if needed
     if (dagContext().isMPPTask())
         /// MPPTask do not need the returned blocks.
-        executeUnion(pipeline, max_streams, dagContext().log, /*ignore_block=*/true);
+        executeUnion(pipeline, max_streams, dagContext().log, /*ignore_block=*/true, "for mpp");
     else
-        executeUnion(pipeline, max_streams, dagContext().log);
+        executeUnion(pipeline, max_streams, dagContext().log, false, "for non mpp");
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/InterpreterDAG.cpp:90]:
```suggestion
        executeUnion(pipeline, max_streams, dagContext().log, /*ignore_block=*/false, "for non mpp");
```

**Review by SeaRise — APPROVED** [review]

**Pull request merged**